nixos/modules/services/audio/upmpdcli.nix

```
@@ -0,0 +1,88 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+with pkgs;
+with lib;
+
+let
+  uid = config.ids.uids.upmpdcli;
+  gid = config.ids.gids.upmpdcli;
+  cfg = config.services.upmpdcli;
+
+  defaultConf = {
+    # upnp network parameters
+    ##upnpiface =
+    upnpip = "127.0.0.1";
+    upnpport = 49152; # libupnp/pupnp defaults to using the first free port after 49152. Note that clients do not need to know about the value, which is automatically discovered.
+
+    ## media renderer parameters
+    friendlyname = "UpMpd";
+    upnpav = 1;
+    openhome = 0;
+    ##lumincompat = 0
+    ##saveohcredentials = 1
+    ##checkcontentformat = 1
+    ##iconpath = /usr/share/upmpdcli/icon.png
+    ##cachedir = /var/cache/upmpdcli
+    ##presentationhtml = /usr/share/upmpdcli/presentation.html
+
+    ## mpd parameters
+    mpdhost = "127.0.0.1";
+    mpdport = 6600;
+    ##mpdpassword =
+    ##ownqueue = 1
+  };
+
+  configuration =
+    defaultConf
+    // {
+      ### https://www.lesbonscomptes.com/upmpdcli/upmpdcli-manual.html#UPMPDCLI-CONFIGURATION
+      pkgdatadir = "${pkgs.upmpdcli}/share/upmpdcli";
+
+    }
+    // cfg.configuration;
+
+  upmpdcliConf = writeText "upmpdcli.conf" (lib.generators.toKeyValue { } configuration);
+in
+{
+
+  options = {
+
+    services.upmpdcli = {
+
+      enable = mkEnableOption "upmpdcli, a music player daemon";
+
+      configuration = mkOption {
+        type = types.attrsOf types.unspecified;
+        default = defaultConf;
+        description = ''
+          Key-value pairs that convey parameters about the configuration
+        '';
+      };
```

This is not really a good way of doing it, because the moment you set `configuration.xyz = "whatever";`, the rest of the defaults will not be included.

How about using freeformAttrs with defaults? (consider adding `format = pkgs.formats.keyValue { };` to the let block up top, then you can use `format.generate` to create the config file as well)

```suggestion
      configuration = mkOption {
        description = ''
          Key-value pairs that convey parameters about the configuration
        '';
        default = { };
        type = types.submodule {
          freeformType = (pkgs.formats.keyValue { }).type;
          options = {
            upnpip = mkOption {
              type = types.str;
              default = "127.0.0.1";
              example = "0.0.0.0";
              description = "The address listening for incoming UPnP connections.";
            };
            
            upnpport = mkOption {
              type = types.port;
              default = 49152;
              description = "The port listening for incoming UPnP connections.";
            };
            
            ...
          };
        };
      };
```